core/src/main/scala/caliban/wrappers/ApolloTracing.scala

```diff
@@ -0,0 +1,187 @@
+package caliban.wrappers
+
+import java.time.{ Instant, ZoneId }
+import java.time.format.DateTimeFormatter
+import java.util.concurrent.TimeUnit
+import caliban.{ GraphQL, Rendering, ResponseValue }
+import caliban.ResponseValue.{ ListValue, ObjectValue }
+import caliban.Value.{ IntValue, StringValue }
+import caliban.wrappers.Wrapper.{ FieldWrapper, OverallWrapper, ParsingWrapper, ValidationWrapper }
+import zio.clock.Clock
+import zio.clock
+import zio.duration.Duration
+import zio.{ FiberRef, UIO }
+import zquery.ZQuery
+
+object ApolloTracing {
+
+  /**
+   * Attaches to the given GraphQL API definition a wrapper that adds tracing information to every response
+   * following Apollo Tracing format: https://github.com/apollographql/apollo-tracing.
+   * @param api a GraphQL API definition
+   */
+  def apolloTracing[R](api: GraphQL[R]): UIO[GraphQL[Clock with R]] =
+    FiberRef
+      .make(Tracing())
+      .map(
+        ref =>
+          api
+            .withWrapper(apolloTracingOverallWrapper(ref))
+            .withWrapper(apolloTracingParsingWrapper(ref))
+            .withWrapper(apolloTracingValidationWrapper(ref))
+            .withWrapper(apolloTracingFieldWrapper(ref))
+      )
+
+  private val dateFormatter: DateTimeFormatter = DateTimeFormatter
+    .ofPattern("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'")
+    .withZone(ZoneId.of("UTC"))
+
+  case class Parsing(startOffset: Long = 0, duration: Duration = Duration.Zero) {
+    def toResponseValue: ResponseValue =
+      ObjectValue(List("startOffset" -> IntValue(startOffset), "duration" -> IntValue(duration.toNanos)))
+  }
+
+  case class Validation(startOffset: Long = 0, duration: Duration = Duration.Zero) {
+    def toResponseValue: ResponseValue =
+      ObjectValue(List("startOffset" -> IntValue(startOffset), "duration" -> IntValue(duration.toNanos)))
+  }
+
+  case class Resolver(
+    path: List[Either[String, Int]] = Nil,
+    parentType: String = "",
+    fieldName: String = "",
+    returnType: String = "",
+    startOffset: Long = 0,
+    duration: Duration = Duration.Zero
+  ) {
+    def toResponseValue: ResponseValue =
+      ObjectValue(
+        List(
+          "path" -> ListValue((Left(fieldName) :: path).reverse.map {
```

Comment:
Why does it need to be reversed here but not when creating the error?

Reply:
Oops, good catch. I will add a test for the error case.